src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReader.cs

```diff
@@ -0,0 +1,227 @@
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Reads byte from TDS stream asynchronously.
+        /// </summary>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="byte"/> value.</returns>
+        public async ValueTask<byte> ReadByteAsync(bool isAsync, CancellationToken ct)
+        {// Throw operation canceled exception before write.
+            ct.ThrowIfCancellationRequested();
+
+            return isAsync ?
+                await _tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false)
+                : (byte)_tdsStream.ReadByte(); // UNSAFE - TODO Expose ReadByte from TdsStream that returns 'byte'.
+        }
+        /// <summary>
+        /// Reads next char from TDS stream asynchronously.
+        /// </summary>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="char"/> value.</returns>
+        public ValueTask<char> ReadCharAsync(bool isAsync, CancellationToken ct)
+            => DoWithRentedBuffer(sizeof(char), async buffer =>
+            {
+                await ReadBytesAsync(buffer, isAsync, ct).ConfigureAwait(false);
+                return (char)((buffer[1] << 8) + buffer[0]);
+            });
+
+        /// <summary>
+        /// Reads char array from Tds Stream of defined <paramref name="length"/> asynchronously.
+        /// </summary>
+        /// <param name="length">Length of array to read.</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="char"/> array.</returns>
+        public async ValueTask<char[]> ReadCharArrayAsync(int length, bool isAsync, CancellationToken ct)
+        {
+            char[] result = new char[length];
+            for (int i = 0; i < length; i++)
+            {
+                result[i] = await ReadCharAsync(isAsync, ct).ConfigureAwait(false);
+                if (!BitConverter.IsLittleEndian)
+                {
+                    result[i] = (char)BinaryPrimitives.ReverseEndianness((ushort)result[i]);
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Reads short value from Tds Stream asynchronously.
+        /// </summary>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token.</param>
+        /// <returns>Async <see cref="ValueTask"/> that returns a <see cref="short"/> value.</returns>
+        public ValueTask<short> ReadInt16Async(bool isAsync, CancellationToken ct)
```

Comment:
Tying to roji and Tornhoof's earlier comments, the rough pseudocode I'd suggest for the strongly-typed read logic is:
```csharp
public ValueTask<X> ReadXAsync(int lengthOfX, bool isAsync, CancellationToken ct)
{
    const int StackAllocationThreshold = 512;

    byte[] rentedBuffer = isAsync || lengthOfX > StackAllocationThreshold
        ? ArrayPool<byte>.Shared.Rent(lengthOfX)
        : null;
    Span<byte> buffer = rentedBuffer == null
        ? stackalloc byte[StackAllocationThreshold].Slice(0, lengthOfX)
        : rentedBuffer.AsSpan(0, lengthOfX);

    try
    {
        ct.ThrowIfCancellationRequested();

        int bytesRead = isAsync
            ? await _tdsStream.ReadAsync(rentedBuffer.AsMemory(0, lengthOfX), ct).ConfigureAwait(false)
            : _tdsStream.Read(buffer);

        if (bytesRead != lengthOfX)
        {
            throw new InvalidOperationException(...);
        }

        return BinaryPrimitives.ReadInt16LittleEndian(buffer);
    }
    finally
    {
        if (rentedBuffer != null)
        {
            ArrayPool<byte>.Shared.Return(rentedBuffer);
        }
    }
}
```

This simplifies when the type `X` and `lengthOfX` is constant, and it's possible that the comparison against `bytesRead` is unnecessary. The main objective is to avoid unnecessary heap allocations, but it makes the use of `TdsBufferManager` more difficult. It also doesn't account for streaming data, or long strings/byte arrays.

Reply:
I have very little context and knowledge of the overall design, but this seems all seems like it would do a ReadAsync for 2 bytes (Int16); at the very least this would mean an async state machine (even on sync completion, where the two bytes are buffered somewhere inside `_tdsStream`), for each and every little read.

For efficient async I/O, it's usually a good idea to split apart the I/O code path - which is the async part where actual I/O can happen, and which is only ever invoked when you know you need to do I/O (not enough bytes cached), and processing/decoding, which operates purely on memory that's known to conain cached data, and is purely synchronous code. For example, this allows you to first make sure that you've read X bytes which you know you're going to need (for a big/complex protocol message or similar), and then you can process that 100-byte chunk efficiently without ever doing async etc (this is how Npgsql works).

Basically the game is how to reduce the amount of needless async/await by lifting that code up the stack, and creating big "chunks" of I/O, rather than letting async permeate everywhere, causing tons of state machines everywhere where you need ro read a byte or two.

One last thing - putting too much pressure on the shared array pool can end up being a source of perf problems too, as contention rises (because too many threads are constantly doing rent/return for tiny buffers). Npgsql has a per-connection private buffer into which we read (so zero contention by design) - you could also have a private ArrayPool (as opposed to the global shared one) on your physical connections. But of course, if/when you know you have a small, fixed-size payload, stackalloc is indeed best.

Reply:
> I have very little context and knowledge of the overall design

This is a very valid point, it is hard to suggest improvements, if we don't really know the invariants,   e.g. which .NET version is minimum etc.

There quite a few high performance concepts to work with buffers in recent .NET versions, incl. PipeReader, ReadOnlySequence etc, some are designed to keep the async path as small as possible, a problem here as shay notes. In one of the other discussions, quite a few nice diagrams were shown to better describe the lay of the land, so to speak.
Maybe this is helpful here too.

Reply:
In most cases for the TDS protocol, I agree - I'd prefer to be able to read the first X bytes, read the token type, read the remaining bytes in the token and interpret the Span/Memory. For most TDS tokens in a token stream (each subsection of [section 2.2.7](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/67b6113c-d722-42d1-902c-3f6e8de09173) in the spec is one type of token) we have a length field, so this is pretty practical.

Reading row data is harder though. As I understand it, SQL Server sends a COLMETADATA token describing the column data types (and in some cases, lengths) in each result set. It then sends a ROW token, followed by a continuous stream of data (the spec refers to this as a tokenless stream). We don't know the length of this ROW token, and can't guarantee that we can calculate it from COLMETADATA; in the case of nullable or dynamic-length data, there's no choice but to read from the TDS stream field-by-field. Once all rows in the result set have been sent, it sends a DONE token which contains the number of rows.

A field is naturally its own type of data - it's essentially the same basic principles and problems...

If COLMETADATA indicates that X consecutive fields are non-nullable and fixed-length then we can optimise the reads a little, and we might even know the minimum (and/or total) length of each row which follows. I think we should read byte Spans and interpret rows outside of the IO completion in this case.

The TDS protocol thus hedges us towards streaming reads. This definitely makes life easier for the SQL Server instance (which can now stream rows and field values as it reads them) but it can make it difficult to separate the IO and interpretation layers. There's a degree of separation here because `TdsReadStream` reads TDS-packet-by-TDS-packet, but it doesn't eliminate the problems which come from allocating the state machines for synchronous completions on the async path.

I'd definitely like to understand how the TDS reader/parser is going to be designed though.

(side note: are you sure that sync completion of the async path allocates? I haven't benchmarked to check, but thought that .NET pooled ValueTask.CompletedTask and that from .NET 5.0 this allocation [didn't occur](https://devblogs.microsoft.com/dotnet/async-valuetask-pooling-in-net-5/#:~:text=When%20used%20as,ValueTask%3CTResult%3E.).)

Reply:
You also need to know that any value longer than a single byte can span packets and you have to deal with it correctly. This sounds easy but forces you to code all read steps knowing that the value may be incomplete and that the minimum number of bytes you will get is 1 and that you may need a network read to get more. This makes life very complicated in the existing codebase. One of the reasons I wanted to use pipelines was that they use ReadOnlySequence which encapsulates this logic very nicely.

The same is true for writes. You are expected to fill every packet apart from the last one in a stream so you must be able to write a minimum of 1 byte of every fixed length value to the output buffer and then write the rest somewhere else. Using the OperationResult pattern you can implement this using retry on too-short buffers rather than using exceptions or having all leaf methods be able to re-request buffers.

Reply:
Addressed majority of the Async-await feedback and simplified back to dedicated buffer for now.
However few follow ups to do:
* Improve buffer allocations (keeping it out of scope here as it may need some refactoring from Streams)
* Redesign reading Char Array and String